File: Sharp.Core/GameEntities/PlayerPawn.cs
/* 
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

using System;
using Sharp.Core.Bridges.Natives;
using Sharp.Core.CStrike;
using Sharp.Core.GameObjects;
using Sharp.Core.Utilities;
using Sharp.Generator;
using Sharp.Shared;
using Sharp.Shared.CStrike;
using Sharp.Shared.Enums;
using Sharp.Shared.GameEntities;
using Sharp.Shared.Types;

namespace Sharp.Core.GameEntities;

internal partial class PlayerPawn : BaseCombatCharacter, IPlayerPawn
{
    public void Print(HudPrintChannel channel,
        string                        message,
        string?                       param1 = null,
        string?                       param2 = null,
        string?                       param3 = null,
        string?                       param4 = null)
        => Player.PawnPrint(_this,
                            channel,
                            channel is HudPrintChannel.Chat && message[0] != '#' ? $" {message}" : message,
                            param1,
                            param2,
                            param3,
                            param4);

    public bool IsPlayer()
        => Player.PawnIsPlayer(_this);

    public void Slay(bool explode = false)
        => Player.PawnSlay(_this, explode);

    public IBaseWeapon? GiveNamedItem(string weapon)
        => BaseWeapon.Create(Player.PawnGiveNamedItem(_this, weapon));

    public IBaseWeapon? GiveNamedItem(EconItemId id)
    {
        if (!Enum.IsDefined(id) || !SharedGameObject.EconItemDefinitionsById.TryGetValue((ushort) id, out var value))
        {
            throw new InvalidOperationException($"Invalid ItemId {id}");
        }

        return GiveNamedItem(value.DefinitionName);
    }

    public IPlayerController? GetController()
        => PlayerController.Create(Player.PawnGetController(_this));

    public IPlayerController? GetOriginalController()
        => PlayerController.Create(OriginalControllerHandle.GetEntityPtr());

    public IPlayerController? GetControllerAuto()
        => IsAlive ? GetController() : GetOriginalController();

    public IBaseWeapon? GetActiveWeapon()
        => BaseWeapon.Create(Player.PawnGetActiveWeapon(_this));

    public IBaseWeapon? GetWeaponBySlot(GearSlot slot, int position = -1)
        => BaseWeapon.Create(Player.PawnGetWeaponBySlot(_this, slot, position));

    public void RemovePlayerItem(IBaseWeapon item)
        => Player.PawnRemovePlayerItem(_this, item.GetAbsPtr());

    public void RemoveAllItems(bool removeSuit = false)
        => Player.PawnRemoveAllItems(_this, removeSuit);

    public void DropWeapon(IBaseWeapon item)
        => Player.PawnDropWeapon(_this, item.GetAbsPtr());

    public bool SelectItem(IBaseWeapon item)
        => Player.PawnSelectItem(_this, item.GetAbsPtr());

    public bool DetachWeapon(IBaseWeapon item)
        => Player.PawnDetachWeapon(_this, item.GetAbsPtr());

    public void SwitchWeapon(IBaseWeapon? weapon)
        => Player.PawnSwitchWeapon(_this, weapon?.GetAbsPtr() ?? nint.Zero);

    public void GiveGloves(int itemDefIndex, int prefab, float wear, int seed)
        => Player.PawnGiveGloves(_this, itemDefIndex, prefab, wear, seed);

    public void GiveGloves(EconGlovesId id, int prefab, float wear, int seed)
    {
        if (!Enum.IsDefined(id))
        {
            Bridges.Natives.Core.LogWarning($"[PlayerPawn::GiveGloves] Invalid GlovesId {id}");

            return;
        }

        Player.PawnGiveGloves(_this, (int) id, prefab, wear, seed);
    }

    public void TransientChangeTeam(CStrikeTeam team)
        => SetTeamLocal(team);

    public void TransientChangeVelocityModifier(float velocityModifier)
        => SetVelocityModifierLocal(velocityModifier);

    public unsafe Vector GetEyeAngles()
        => *Entity.GetEyeAngles(_this);

    public unsafe Vector GetEyePosition()
        => *Entity.GetEyePosition(_this);

    public unsafe SoundOpEventGuid EmitSoundClient(string sound, float? volume = null)
        => Player.PawnEmitSoundClient(_this, sound, &volume);

#region Schemas

    [NativeSchemaField("CCSPlayerPawn", "m_ArmorValue", typeof(int))]
    private partial SchemaField GetArmorValueField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_flFlashMaxAlpha", typeof(float))]
    private partial SchemaField GetFlashMaxAlphaField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_flFlashDuration", typeof(float))]
    private partial SchemaField GetFlashDurationField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_iPlayerState", typeof(PlayerState))]
    private partial SchemaField GetStateField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_iNumSpawns", typeof(int))]
    private partial SchemaField GetNumSpawnsField();

    [NativeSchemaField("CCSPlayerPawn", "m_fMolotovDamageTime", typeof(float))]
    private partial SchemaField GetMolotovDamageTimeField();

    [NativeSchemaField("CBasePlayerPawn", "m_pItemServices", typeof(ItemService))]
    private partial SchemaField GetItemServiceField();

    [NativeSchemaField("CBasePlayerPawn", "m_pMovementServices", typeof(MovementService))]
    private partial SchemaField GetMovementServiceField();

    [NativeSchemaField("CBasePlayerPawn", "m_pWeaponServices", typeof(WeaponService))]
    private partial SchemaField GetWeaponServiceField();

    [NativeSchemaField("CBasePlayerPawn", "m_pCameraServices", typeof(CameraService))]
    private partial SchemaField GetCameraServiceField();

    [NativeSchemaField("CBasePlayerPawn", "m_pObserverServices", typeof(ObserverService))]
    private partial SchemaField GetObserverServiceField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_hOriginalController", typeof(CEntityHandle<IPlayerController>))]
    private partial SchemaField GetOriginalControllerHandleField();

    [NativeSchemaField("CCSPlayerPawn", "m_flHealthShotBoostExpirationTime", typeof(float))]
    private partial SchemaField GetHealthShotBoostExpirationTimeField();

    [NativeSchemaField("CCSPlayerPawn", "m_EconGloves", typeof(EconItemView), InlineObject = true)]
    private partial SchemaField GetEconGlovesField();

    [NativeSchemaField("CCSPlayerPawn", "m_flVelocityModifier", typeof(float))]
    private partial SchemaField GetVelocityModifierField();

    [NativeSchemaField("CCSPlayerPawn", "m_iShotsFired", typeof(int))]
    private partial SchemaField GetShotsFiredField();

    [NativeSchemaField("CCSPlayerPawn", "m_flFlinchStack", typeof(float))]
    private partial SchemaField GetFlinchStackField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_flProgressBarStartTime", typeof(float))]
    private partial SchemaField GetProgressBarStartTimeField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_iProgressBarDuration", typeof(int))]
    private partial SchemaField GetProgressBarDurationField();

    [NativeSchemaField("CBasePlayerPawn", "m_iHideHUD", typeof(uint))]
    private partial SchemaField GetHideHudField();

    [NativeSchemaField("CBasePlayerPawn", "m_fTimeLastHurt", typeof(float))]
    private partial SchemaField GetTimeLastHurtField();

    [NativeSchemaField("CBasePlayerPawn", "m_flDeathTime", typeof(float))]
    private partial SchemaField GetDeathTimeField();

    [NativeSchemaField("CBasePlayerPawn", "m_fNextSuicideTime", typeof(float))]
    private partial SchemaField GetNextSuicideTimeField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_bRespawning", typeof(bool))]
    private partial SchemaField GetRespawningField();

    [NativeSchemaField("CCSPlayerPawnBase", "m_fNextRadarUpdateTime", typeof(float))]
    private partial SchemaField GetNextRadarUpdateTimeField();

    [NativeSchemaField("CCSPlayerPawn", "m_bInBuyZone", typeof(bool))]
    private partial SchemaField GetInBuyZoneField();

    [NativeSchemaField("CCSPlayerPawn", "m_bInHostageRescueZone", typeof(bool))]
    private partial SchemaField GetInHostageRescueZoneField();

    [NativeSchemaField("CCSPlayerPawn", "m_bInBombZone", typeof(bool))]
    private partial SchemaField GetInBombZoneField();

    [NativeSchemaField("CCSPlayerPawn", "m_bIsBuyMenuOpen", typeof(bool))]
    private partial SchemaField GetIsBuyMenuOpenField();

    [NativeSchemaField("CCSPlayerPawn", "m_bLeftHanded", typeof(bool))]
    private partial SchemaField GetLeftHandedField();

    [NativeSchemaField("CCSPlayerPawn", "m_fSwitchedHandednessTime", typeof(float))]
    private partial SchemaField GetSwitchedHandednessTimeField();

    [NativeSchemaField("CCSPlayerPawn", "m_vecTotalBulletForce", typeof(Vector))]
    private partial SchemaField GetTotalBulletForceField();

    [NativeSchemaField("CCSPlayerPawn", "m_iDeathFlags", typeof(int))]
    private partial SchemaField GetDeathFlagsField();

    [NativeSchemaField("CCSPlayerPawn", "m_bWaitForNoAttack", typeof(bool))]
    private partial SchemaField GetWaitForNoAttackField();

    [NativeSchemaField("CCSPlayerPawn", "m_aimPunchAngle", typeof(Vector))]
    private partial SchemaField GetAimPunchAngleField();

    [NativeSchemaField("CCSPlayerPawn", "m_aimPunchAngleVel", typeof(Vector))]
    private partial SchemaField GetAimPunchAngleVelocityField();

    [NativeSchemaField("CCSPlayerPawn", "m_aimPunchTickBase", typeof(int))]
    private partial SchemaField GetAimPunchTickBaseField();

    [NativeSchemaField("CCSPlayerPawn", "m_aimPunchTickFraction", typeof(float))]
    private partial SchemaField GetAimPunchTickFractionField();

    [NativeSchemaField("CCSPlayerPawn",
                       "m_aimPunchCache",
                       typeof(SchemaUnmanagedVector<Vector>),
                       InlineObject = true,
                       ReturnType = typeof(ISchemaList<Vector>))]
    private partial SchemaField GetAimPunchCacheField();

#endregion
}

File: Engine/src/cstrike/entity/PlayerPawn.cpp
/* 
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

#include "gamedata.h"
#include "global.h"
#include "vhook/call.h"

#include "cstrike/component/PlayerPawnComponent.h"
#include "cstrike/entity/CBaseWeapon.h"
#include "cstrike/entity/PlayerController.h"
#include "cstrike/entity/PlayerPawn.h"
#include "cstrike/interface/CGameEntitySystem.h"
#include "cstrike/type/CEconItemView.h"
#include "cstrike/type/EmitSound.h"

#include <cstdarg>

CBasePlayerController* CBasePlayerPawn::GetControllerInternal()
{
    return g_pGameEntitySystem->FindEntityByEHandle<CBasePlayerController*>(m_hController());
}

CBaseWeapon* CBasePlayerPawn::GiveNamedItem(const char* pszWeaponName)
{
    const auto pItemServices = m_pItemServices();
    if (!pItemServices)
        return nullptr;
    return pItemServices->GiveNamedItem(pszWeaponName);
}

void CBasePlayerPawn::RemoveAllItems(bool removeSuit)
{
    const auto pItemServices = m_pItemServices();
    if (!pItemServices)
        return;
    return pItemServices->RemoveAllItems(removeSuit);
}

CBasePlayerPawn* CBasePlayerPawn::FindBySlot(PlayerSlot_t slot)
{
    if (const auto pController = CBasePlayerController::FindBySlot(slot))
        return g_pGameEntitySystem->FindEntityByEHandle<CBasePlayerPawn*>(pController->m_hPawn());

    return nullptr;
}

void CBasePlayerPawn::ClientPrint(HudPrint_t dest, const char* name, const char* param1, const char* param2, const char* param3, const char* param4)
{
    const auto pController = GetController<CBasePlayerController*>();
    if (!pController)
        return;

    pController->ClientPrint(dest, name, param1, param2, param3, param4);
}

void CBasePlayerPawn::ClientPrintF(HudPrint_t dest, const char* message, ...)
{
    va_list args;
    va_start(args, message);

    char buffer[256];
    vsnprintf(buffer, sizeof(buffer), message, args);

    va_end(args);

    ClientPrint(dest, buffer);
}

void CBasePlayerPawn::CommitSuicide(bool explode)
{
    VCall_AutoVoid(CBasePlayerPawn, CommitSuicide, this, explode, true);
}

CBaseWeapon* CBasePlayerPawn::GetActiveWeapon()
{
    const auto pWeaponServices = m_pWeaponServices();
    AssertPtr(pWeaponServices);

    return g_pGameEntitySystem->FindEntityByEHandle<CBaseWeapon*>(pWeaponServices->m_hActiveWeapon());
}

CBaseWeapon* CBasePlayerPawn::GetWeaponBySlot(GearSlot_t slot, int32_t position)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return nullptr;

    return pWeaponServices->GetWeaponBySlot(slot, position);
}

void CBasePlayerPawn::RemovePlayerItem(CBaseWeapon* pItem)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return;

    pWeaponServices->RemovePlayerItem(pItem);
}

bool CBasePlayerPawn::DetachWeapon(CBaseWeapon* pWeapon)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return false;

    return pWeaponServices->DetachWeapon(pWeapon);
}

void CBasePlayerPawn::DropWeapon(CBaseWeapon* pWeapon)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return;

    pWeaponServices->DropWeapon(pWeapon, nullptr, nullptr);
}

bool CBasePlayerPawn::SelectItem(CBaseWeapon* pWeapon)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return false;

    return pWeaponServices->SelectItem(pWeapon);
}

void CBasePlayerPawn::SwitchWeapon(CBaseWeapon* pWeapon)
{
    const auto pWeaponServices = m_pWeaponServices();
    if (!pWeaponServices)
        return;

    pWeaponServices->SwitchWeapon(pWeapon);
}

void CBasePlayerPawn::TransientChangeTeam(CStrikeTeam_t team) const
{
    static auto offset = GetFieldOffset("CBaseEntity", "m_iTeamNum");
    SetFieldValue(offset, team);
}

SndOpEventGuid_t CBasePlayerPawn::EmitSoundClient(const char* pszSound, const float* pVolume)
{
    const auto pController = GetController<CBasePlayerController*>();
    if (!pController)
    {
        FERROR("EmitSoundClient with nullptr controller -> sound [%s] ", pszSound);
        return g_DefaultSndOpEventGuid;
    }

    return pController->EmitSoundClient(pszSound, pVolume);
}

void CCSPlayerPawn::GiveGloves(int itemDefIndex, int prefab, float wear, int seed)
{
    AssertBool(itemDefIndex >= 0 && itemDefIndex <= 65535);

    const auto pItem = m_EconGloves();
    if (const auto attributes = pItem->m_NetworkedDynamicAttributes())
    {
        pItem->m_iItemDefinitionIndex(static_cast<uint16_t>(itemDefIndex));
        attributes->SetOrAddAttributeValueByName("set item texture prefab", static_cast<float>(prefab));
        attributes->SetOrAddAttributeValueByName("set item texture wear", static_cast<float>(wear));
        attributes->SetOrAddAttributeValueByName("set item texture seed", static_cast<float>(seed));

        pItem->m_iItemIDLow(-1);
        pItem->m_bInitialized(true);

        SetBodyGroupByName("default_gloves", 1);
    }
}

CBasePlayerController* CCSPlayerPawnBase::GetOriginalControllerInternal()
{
    return g_pGameEntitySystem->FindEntityByEHandle<CBasePlayerController*>(m_hOriginalController());
}

bool CCSPlayerPawnBase::IsPlayer()
{
    static auto offset = g_pGameData->GetVFunctionIndex("CCSPlayerPawnBase::IsPlayer");
    return CALL_VIRTUAL(bool, offset, this);
}
